R/sequential.R

```diff
@@ -0,0 +1,48 @@
+#' Perform multiple requests in sequence
+#'
+#' Given a list of requests, this function requests each one turn, returning
+#' a list of responses.
+#'
+#' @inheritParams req_perform_sequential
+#' @inheritParams req_perform_iteratively
+#' @export
+#' @examples
+#' req <- request("https://api.restful-api.dev/objects")
+#' ids <- sort(sample(100, 50))
+#' chunks <- split(ids, (seq_along(ids) - 1) %/% (length(ids) / 10))
+#' reqs <- chunks %>% lapply(\(idx) req %>% req_url_query(id = idx, .multi = "comma"))
+#' resps <- reqs %>% req_perform_sequential()
+#' resps_data(resps, \(resp) resp_body_json(resp))
+req_perform_sequential <- function(reqs, paths = NULL, progress = TRUE) {
+  if (!is.list(reqs)) {
+    stop_input_type(reqs, "a list")
+  }
+  if (!is.null(paths)) {
+    if (length(reqs) != length(paths)) {
+      cli::cli_abort("If supplied, {.arg paths} must be the same length as {.arg req}.")
+    }
+  }
+
+  progress <- create_progress_bar(
+    total = length(reqs),
+    name = "Iterating",
+    config = progress
+  )
+
+  resps <- rep_along(reqs, list())
+
+  tryCatch({
+    for (i in seq_along(reqs)) {
+      check_request(reqs[[i]], arg = glue::glue("req[[{i}]]"))
+      resps[[i]] <- req_perform(reqs[[i]], path = paths[[i]])
+      progress$update()
+    }
+  }, interrupt = function(cnd) {
+    cli::cli_alert_warning(
```

Comment:
When is the interrupt actually caught? I always seem to get

```
Error in `req_perform()` at httr2/R/sequential.R:37:6:  1s
! Failed to perform HTTP request.
Caused by error in `curl::curl_fetch_memory()`:
! Operation was aborted by an application callback
```

when I interrupt the example with the escape key.

Reply:
Oh shoot, I somehow never hit that when I tested this before, but now it's all I see 😢

Reply:
And I already filed an issue on this in curl: https://github.com/jeroen/curl/issues/271. I guess I'll leave this code here since it will eventually be useful but there's not much we can do until then.